### PR TITLE
Add ARCHITECTURE.md

### PR DESCRIPTION
**Pull request by hlopko**
In this file I hope to describe how we think about the architecture
of `rules_rust`. It's goal is to help contributors orient themselves, and to
document code restrictions and assumptions.

### CODE REVIEW

**Comment by hlopko** [comment]
Also r? @UebelAndre :)

**Comment by hlopko** [comment]
Comments and contributions are very welcome! This is just a first draft. I haven't written anything about many packages such as proto or bindgen.

**Comment by UebelAndre** [comment]
😄 Thanks for starting this! I think it looks good so far.

Does this mean you're looking to move `./util` into `./rust/tools`?

I'm also particularly curious about `cargo`. I think I like it as it's own package but I can also see it as part of `rust/tools`. Your thoughts?

**Review by dfreese — APPROVED** [review]

**Comment by dfreese** [comment]
Looks like a solid start.  Thanks for putting that together.

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+depend on `//rust` and its subpackages. Exceptions are Bazel's builtin packages
+and public APIs of other rules (such as `rules_cc`). Depending on
+`//rust/private` from packages other than `//rust` is not supported, we reserve
+the right to change anything there and not tell anybody.
+
+When core rules need custom tools (such as process wrapper, launcher, test
+runner, and so on), they should be stored in `//rust/tools` (for public tools)
+or in `//rust/private/tools` (for private tools). These should:
+
+* essential (tools that are not essential can be used conditionally for example
```
Comment (illicitonion) [inline ARCHITECTURE.md:37]: I'm not quite sure what this bullet point is trying to tell me?

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+This is the core package of the rules. It contains all the core rules such as
+`rust_binary` and `rust_library`. It also contains `rust_common`, a Starlark
+struct providing all rules_rust APIs that one might need when writing custom
+rules integrating with rules_rust.
+
+`//rust` and all its subpackages have to be standalone. Users who only need
+the core rules should be able to ignore all other packages.
+
+`//rust:defs.bzl` is the file that all users of `rules_rust` will be using.
+Everything in this file can be used (dependend on) and is supported. Typically
```
Comment (illicitonion) [inline ARCHITECTURE.md:21]: This somewhat raises the question of compatibility guarantees - "can be used and is supported"... Until when?

**Review by illicitonion — APPROVED** [review]
Looks good!

**Comment by hlopko** [comment]
> Does this mean you're looking to move ./util into ./rust/tools?

Yup (or //rust/private/tools, depending on the tool, I haven't thought about where tools should go).

> I'm also particularly curious about cargo. I think I like it as it's own package but I can also see it as part of rust/tools. Your thoughts?

If you're asking about cargo build script runner, I'm open to suggestions as I haven't looked into how it is used and how the rule using it works. If there's not strong reason why it should be in //rust, I'd vote to not put it there.

File: ARCHITECTURE.md (hunk 1 above)
Comment (hlopko) [reply ARCHITECTURE.md]: Rewrote to say `be essential (it does not make sense to have core rules without them)`. What I was trying to say before was that when core rules can be used both with and without the tool, the tool is not essential and therefore should probably not be bundled in the core rules.

File: ARCHITECTURE.md (hunk 2 above)
Comment (hlopko) [reply ARCHITECTURE.md:21]: That's a hard question which I'd like to find answer to in a separate discussion. Would you prefer uncertainty of the current text, or deleting it?

**Comment by hlopko** [comment]
So right now we say //rust is core package, everything else is less so. One alternative is to have a //contrib package that would contain all the "less so" packages. Opinions?

**Comment by UebelAndre** [comment]
> If you're asking about cargo build script runner, I'm open to suggestions as I haven't looked into how it is used and how the rule using it works. If there's not strong reason why it should be in //rust, I'd vote to not put it there.

To my knowledge, the build script runner is there to run [`build.rs` scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html) which is a concept in Cargo, not rustc. So I think it makes more sense to leave it in the `cargo` directory. I think I like the existence of a `cargo` directory because it can act as the separation of `cargo` and `rustc`. Notably, https://github.com/bazelbuild/rules_rust/pull/598 and https://github.com/bazelbuild/rules_rust/pull/590 are also something I feel should go into `cargo` since they build on cargo manifests to generate Bazel repository definitions.

File: ARCHITECTURE.md (hunk 2 above)
Comment (illicitonion) [reply ARCHITECTURE.md:21]: I'd suggest either adding an explicit qualifier ("though stability is not currently guaranteed across commits", and a link to an issue), or removing it.

**Review by damienmg — APPROVED** [review]

**Comment by damienmg** [comment]
> > If you're asking about cargo build script runner, I'm open to suggestions as I haven't looked into how it is used and how the rule using it works. If there's not strong reason why it should be in //rust, I'd vote to not put it there.
> 
> To my knowledge, the build script runner is there to run [`build.rs` scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html) which is a concept in Cargo, not rustc. So I think it makes more sense to leave it in the `cargo` directory. I think I like the existence of a `cargo` directory because it can act as the separation of `cargo` and `rustc`. Notably, #598 and #590 are also something I feel should go into `cargo` since they build on cargo manifests to generate Bazel repository definitions.

Yes `cargo_build_script` is named like this and in a different package because it is an interface with cargo not with the rust toolchain.

File: ARCHITECTURE.md (hunk 2 above)
Comment (hlopko) [reply ARCHITECTURE.md:21]: Good point. Done.

**Comment by hlopko** [comment]
I've created https://github.com/bazelbuild/rules_rust/issues/601 and https://github.com/bazelbuild/rules_rust/issues/601 to track cleanups relevant to this PR.

**Pull request merged**